---
title: "Codon Positions"
output: html_notebook
---
Tung Nguyen
November 21, 2022

# Initialize samples & variant-level data: samples_n_var, minor_variant_sites_threshold 
 
samples: samples_n_var
variant: minor_variant_sites_threshold

```{r}
# Load common functions, libraries, and -- data/variables from 01_load_new_samples data 
source("./scripts/startup.R")
```

# Compute positions of codons


```{r}
var = read_feather('processing/minor_variants_filtered_100x0.01_50.arrow')
samples_n_var = read_feather("processing/samples_n_var.arrow")

# first iteration, just in small intervals
genes = fread("ntpos_gene_update.csv", data.table = F);
codon_translate = fread("codon.txt", data.table=F, sep = "\t", skip = 2, fill = T) %>% 
    select(ref_codon = 1, ref_sym = 3);

gene_start_end = genes %>% group_by(gene_id) %>% summarize(start = min(ntpos), 
                                                           end = max(ntpos)) %>% 
  arrange(start) %>% as.data.frame

for (i in 1:nrow(gene_start_end)) {
    gene = gene_start_end[i, "gene_id"] %>% .[[1]]
    assign_gene = (var$ntpos >= gene_start_end[i, "start"] & var$ntpos <= gene_start_end[i, "end"])
    var[assign_gene, "gene"] = gene
    var[assign_gene, "codon_pos"] = ((var$ntpos[assign_gene] - gene_start_end[i, "start"]) %% 3) + 1
    var[assign_gene, "aapos"] = ceiling((var$ntpos[assign_gene] - gene_start_end[i, "start"] + 1) / 3)
    var[assign_gene,"ntpos_codonstart"] = var$ntpos[assign_gene]+1-var$codon_pos[assign_gene]
}

    # if pos 1, get the next two basepairs
    # x, x+1, x+2

    # if pos 2, get the previous and the next one
    # x-1, x, x+1

    # if pos 3, get the previous 
    # x-2, x-1, x

    # range rule: (x+1-pos):(x+1-pos)+2

for (i in 1:nrow(gene_start_end)) {
    gene = gene_start_end[i, "gene_id"] %>% .[[1]]
    assign_gene = (genes$ntpos >= gene_start_end[i, "start"] & genes$ntpos <= gene_start_end[i, "end"])
    genes[assign_gene, "codon_pos"] = ((genes$ntpos[assign_gene] - gene_start_end[i, "start"]) %% 3) + 1
    genes[assign_gene, "aapos"] = ceiling((genes$ntpos[assign_gene] - gene_start_end[i, "start"] + 1) / 3)
}

aa = genes %>% group_by(gene_id, aapos) %>% 
  summarize(ref_codon = paste(refnt,collapse=","), ntpos_codonstart = min(ntpos)) %>% 
  separate(ref_codon, sep = ",", into = c("ref.1", "ref.2", "ref.3"), remove = F) %>% 
  mutate(ref_codon = gsub(",","",ref_codon)) %>%
  filter(!gene_id=="") %>% left_join(codon_translate)

var_aa = var %>% left_join(aa)

# if codon pos is #1
major_string = data.frame(first = c("major", "ref.2", "ref.3"), 
                 second = c("ref.1", "major", "ref.3"), 
                 third = c("ref.1", "ref.2", "major"));
minor_string = major_string %>% mutate( across(.cols = everything(), ~str_replace(., "major", "minor")))
# major flaw in the above is that it assumes there are no more than 1 mutation 
# (minor or major) in the same codon. For us it does not matter, because we are doing codon analysis
# involving just positional importance not actual amino acid conversion analysis. Thus the above is
# just rough codon sequence analysis.

for (i in 1:3) {
    idx = (var_aa$codon_pos == i)
    var_aa[idx, "majorcodon"] = var_aa %>% filter(codon_pos == i) %>% 
           unite("tmp", major_string[,i], sep = "") %>% select(tmp)
    var_aa[idx, "minorcodon"] = var_aa %>% filter(codon_pos == i) %>% 
            unite("tmp", minor_string[,i], sep = "") %>% select(tmp)
}

# translate the major and minor variants
patient_data <- fread("sample_and_patient_data.csv",data.table=F) %>% 
  mutate(MCoVNumber=str_remove(mcov_id, "-")) %>% mutate(collection_date=as.Date(COLLECTION_DT, "%m/%d/%y")) %>% 
  mutate(collection_month=format(as.Date(collection_date), "%Y-%m")) %>%
  mutate(CT=ifelse(INSTRUMENT_RESULT<50, INSTRUMENT_RESULT, NA_integer_)) %>% 
  mutate(vaccine_status=if_else(Vaccine_Status=="No vaccine",0,1)) %>%
  mutate(age18under=if_else(Age_Group=="00-17",1,0)) %>% 
  mutate(age18to54=if_else(Age_Group=="18-54",1,0)) %>% 
  mutate(age55plus=if_else(Age_Group=="55-64"|Age_Group=="65+",1,0)) %>%
  select(MCoVNumber, collection_date, collection_month, run=run_group, CT, 
         ordering_clinic=ORDERING_CLINIC_TYPE,pui=PUI, age18under, age18to54, 
         age55plus, sex=SEX, ethnicity=Ethnicity, obesity=Obesity_YN, 
         chronic_lung_disease=Chronic_Lung_Disease_YN, 
         chronic_liver_disease=Chronic_Liver_Disease_YN, 
         hcw=IS_SURVEILLANCE, chronic_heart_disease=Chronic_Heart_Disease_YN,
         chronic_kidney_disease=Chronic_Kidney_Disease_YN, 
         hypertension=Hypertension_YN, diabetes=Diabetes_YN, 
         cancer=Cancer_YN, hiv=HIV_YN, transplant_patient=Transplant_Patient, 
         vaccine_status, admitted_hospital=Admitted_YN, highest_level=HIGHEST_LEVEL_OF_CARE, 
         max_respiratory_support=MaxRespiratorySupport, mAb=mAb_YN, plasma=Plasma_YN) %>%
  mutate(surveillance = if_else(hcw == "Yes Surveillance",1,0))

factor_columns <- c("collection_month","run","ordering_clinic", "pui", "age18under", 
                    "age18to54", "age55plus","sex","ethnicity","obesity","surveillance", 
                    "chronic_lung_disease","chronic_liver_disease","chronic_heart_disease",
                    "chronic_kidney_disease","hypertension","diabetes","cancer","hiv",
                    "transplant_patient","vaccine_status","admitted_hospital",
                    "highest_level","max_respiratory_support","mAb","plasma") 

patient_data[factor_columns] <- lapply(patient_data[factor_columns], factor)

sample_type = fread("sample_type_PUI.csv", data.table = F) %>% 
  mutate(MCoVNumber=str_remove(mcov_id, "-"))
sample_duration =  fread("timestamp_sample_RNA_extraction_processing.csv",
                        data.table = F) %>% 
  mutate(MCoVNumber=str_remove(mcov_id, "-"))

patient_data[factor_columns]<-lapply(patient_data[factor_columns], factor)

patient_var_tmp = var_aa %>% left_join(codon_translate %>% 
                                         select(majorcodon = ref_codon, tmp_sym = ref_sym)) %>% 
  mutate(majoraa = tmp_sym) %>% select(!tmp_sym) %>% 
  select(!c(gene_id, ref.1, ref.2, ref.3)) %>% 
  full_join(samples_n_var) %>% left_join(patient_data) %>%
 select(-pui) %>% left_join(sample_type) %>% 
  left_join(sample_duration %>% select(-COLLECTION_DT)) %>%
  mutate(vocAlpha=if_else(str_starts(scorpio_call, "Alpha"),1,0), 
         vocDelta=if_else(str_starts(scorpio_call, "Delta"),1,0)) %>% 
  mutate(vocAlpha=as.factor(vocAlpha), vocDelta=as.factor(vocDelta))

#### ACTUAL PATIENT ANALYSIS FOR 02 RMD.
write_feather(patient_var_tmp, "processing/patient_var_tmp.arrow")
patient_var = patient_var_tmp %>% filter(INSTRUMENT_RESULT < 26)
#fwrite(var_aa, "processing/var_aa_ct.txt", sep = "\t", quote = F)
write_feather(patient_var, "processing/patient_var.arrow")
patient_var_30 = patient_var %>% filter(n_var < 30)
write_feather(patient_var_30, "processing/patient_var_30.arrow")
patient_counts = patient_var %>% select(MCoVNumber,lineage:last_col()) %>% unique
write_feather(patient_counts, "processing/patient_counts.arrow")
patient_counts_30 = patient_counts %>% filter(n_var < 30)
write_feather(patient_counts_30, "processing/patient_counts_30.arrow")
```


# Number of counts that are acceptable
```{r}
patient_var_tmp = read_feather("processing/patient_var_tmp.arrow")

idx = seq(0,100,5)
for (i in 2:length(idx)) {
    minorcounts = patient_var_tmp %>% select(name) %>%
    group_by(name) %>% 
    summarize(counts = n()) %>% filter((counts > idx[i-1]) & (counts < idx[i])) %>% 
    as.data.frame()

    minorallele = patient_var_tmp %>% filter((name %in% minorcounts$name)) %>% mutate(codon_pos = as.factor(codon_pos))

    ##### BIN MINOR ALLELE FREQUENCIES (MAF)
    binned_minorallele = minorallele %>% mutate(allele_freq_bin = 
        cut(minorfreq, breaks=seq(0.00, 1, by = 0.05))) %>% 
        drop_na(allele_freq_bin)

    freq_minorallele = binned_minorallele %>% group_by(codon_pos, allele_freq_bin) %>% summarize(counts = n()) %>%
        group_by(allele_freq_bin) %>% 
        left_join(   
            binned_minorallele %>% group_by(allele_freq_bin) %>% summarize(total_bin = n()) # doing the above w/o codon pos
        ) %>% mutate(proportion = counts / total_bin) %>% mutate(n_bin = idx[i])
    
    if (i == 2) {
        saved_freq_minorallele = freq_minorallele
    } else {
        saved_freq_minorallele = rbind(saved_freq_minorallele, freq_minorallele)
    }
}

saved_freq_minorallele_lowestfreq = saved_freq_minorallele %>% filter(allele_freq_bin == "(0,0.05]");

ggline(saved_freq_minorallele_lowestfreq, x = "n_bin", y = "proportion", color = "codon_pos", 
       shape = "codon_pos", palette = "jco") +
        #scale_color_manual(values = c("red", "green", "blue")) + 
    rotate_x_text(90) + ylim(0.2,0.5) + geom_hline(yintercept = 0.33, color = "red",
                                                 linetype = "dashed")

saved_freq_minorallele_lowestfreq

```

# CT cutoff bins that are acceptable
```{r}

input = patient_var_tmp %>% filter(n_var > 0)

ctplot = function(input, xlab = T, legend = T) {
    ct_cutoff = input %>% 
      mutate(codon_pos = as.factor(codon_pos)) %>% 
    filter(INSTRUMENT_RESULT < 40) %>%
    mutate(ct_bin = cut(INSTRUMENT_RESULT, breaks=seq(1, 40, by = 1), right = FALSE)) %>%
    group_by(ct_bin, codon_pos) %>%
    summarise(n = n()) %>% group_by(ct_bin) %>%
    mutate(ct_right = gsub(".*[,]([^.]+)[)].*", "\\1", 
                           as.character(ct_bin)) %>% as.numeric) %>%
    mutate(total = sum(n), prop = n/total) 
    
  
  p = ggplot(ct_cutoff, aes(x = ct_right, y = prop, 
                            shape = codon_pos, color = codon_pos)) + 
    theme_pubr() + geom_line() + geom_point() +
    geom_hline(yintercept = 0.33, color = "grey", linetype = "dashed") +
    geom_vline(xintercept = seq(6,40,2), color = "grey", linetype = "dotted") +
    geom_vline(xintercept = 26, color = "grey10", linetype = "dashed") +
    scale_y_continuous(limits = c(.2,.5)) +
    scale_x_continuous(breaks=seq(0,40,10))
  
  if (xlab == F) {
    p = p + xlab(NULL) + ylab(NULL) +
    scale_x_continuous(breaks=seq(0,40,10))
  }
  if (legend == F) {
    p = p + theme(legend.position='none')
  }
  
  xhist = 
    axis_canvas(p, axis = "x") + 
    geom_col(data = ct_cutoff %>% select(ct_bin, ct_right, total) %>% distinct(),
                   aes(x = ct_right, y = total),
                   color = 'grey', stat = "identity")
  out_plot = p %>%
    insert_xaxis_grob(xhist, grid::unit(0.5, "in"), position = "top") %>%
    ggdraw()
}

nall = ctplot(input)
print(nall)

```

Different quadrants
```{r}

n30plus = ctplot(input %>% filter(n_var > 30), legend = F, xlab = F)
n30minus = ctplot(input %>% filter(n_var < 30), legend = F, xlab = F) + theme(legend.position = "none")
#plot_grid(n30plus, n30minus, nrow = 2, align = "v")

plot1_codon_counts = plot_grid(n30plus,
          n30minus, 
          nrow = 2, ncol = 1, 
          align = "hv")

plot1_codon_counts

ggsave("ggsave/plot1_codon_counts.pdf", plot1_codon_counts, height = 4, width = 4)
```
# CODON BIAS EDA
More codon bias stuff
```{r}
idx = seq(0,50,5)
nbin = function(patient_var, MAF = 0.01, idx = seq(0,50,5)) {
  tmp = patient_var %>%
    filter(minorfreq > MAF) %>%
    mutate(n_bin = cut(n_var, breaks = idx)) %>% 
    group_by(n_bin, codon_pos) %>%
    summarize(counts = n())
  totals_bin = tmp %>% group_by(n_bin) %>% summarize(total = sum(counts))
  tmp = tmp %>% left_join(totals_bin) %>% mutate(proportion = counts / total) %>%
    mutate(codon_pos = as.factor(codon_pos)) %>% 
    mutate(n_threshold = gsub(".*[,]([^.]+)[]].*", "\\1", 
                           as.character(n_bin)) %>% as.numeric)
  tmp$MAF = MAF
  return(tmp)
}
```

```{r}
# MAF_range = c(0.01,0.02,0.05,0.10,0.2)
# index = 0
# for (MAF in MAF_range) {
#   index = index + 1
#   tmp = nbin(patient_var_tmp, MAF, idx = seq(0,50,5))
#   if (index == 1) {
#     n_MAF_out = tmp
#   } else {
#     n_MAF_out = rbind(n_MAF_out, tmp)
#   }
# }
# 
# n_MAF_hm = n_MAF_out %>% ungroup %>% filter(codon_pos == 3) %>% 
#   select(n_threshold, proportion, MAF, counts) %>% replace(is.na(.), 55)
# 
# 
# # ggplot(n_MAF_hm, aes(n_threshold, MAF, fill=residual)) + 
# #   geom_tile() + theme_pubr() + scale_fill_viridis(option="A")
# 
# p = ggplot(n_MAF_hm, aes(x = n_threshold, y = proportion, 
#                           color = as.factor(MAF))) + 
#   theme_pubr() + geom_line() + geom_point() + ylim(0.25,0.5) + 
#   geom_hline(yintercept = 0.33, color = "grey10", linetype = "dashed") +
#   geom_vline(xintercept = idx[-1], color = "grey", linetype = "dotted") +
#   scale_x_continuous(breaks=seq(5,50,5), limits= c(5,55)) + 
#   theme(axis.text.x = element_text(angle=90))
# 
# xhist = 
#   axis_canvas(p, axis = "x") + 
#   geom_col(data = n_MAF_hm,
#                  aes(x = n_threshold, y = counts, fill = as.factor(MAF)),
#                 stat = "identity") +
#   geom_text(data = n_MAF_hm %>% 
#               select(MAF, n_threshold, counts) %>% 
#               filter(MAF == MAF_range[1]),
#             aes(x = n_threshold, y = 5000, label = counts, angle = 90))
# 
# p %>%
#   insert_xaxis_grob(xhist, grid::unit(0.5, "in"), position = "top") %>%
#   ggdraw()
```

Generate a fraction of position 3 per sample
```{r}
p_data = patient_var_tmp %>% group_by(MCoVNumber) %>% 
  summarize(n_var, prop3 = sum(codon_pos==3)/n_var) %>% distinct() #%>% 
#  replace_na(list(prop3 = 0))

p = ggplot(p_data, aes(x = n_var, y = prop3)) + 
  geom_point(shape = 21, alpha = 0.1) +
  geom_hline(yintercept = 0.33, linetype = "dashed", color = "grey") + 
  theme_pubr() + scale_x_continuous(trans = "log1p", 
                                    breaks = c(0,1,5,10,50,100,500)) +
  geom_smooth(data = p_data %>% 
                filter(n_var > 0), se = F, 
              aes(x=n_var, y = prop3), method = "loess",
              color = "red")

hist = ggplot(data = patient_counts, aes(x = n_var)) +
  geom_histogram() + 
    geom_violin(aes(y = 3), draw_quantiles = c(0.05,0.1, .25,.5,.75, .90, 0.95), width = 2, 
                  color = "red", alpha = .5) +
  scale_y_continuous(trans = "log1p", breaks = c(0,1,10,100,1000,5000)) +
  scale_x_continuous(trans = "log1p", breaks = c(0,1,5,10,30,100,500)) +
  theme_pubr()

quantile(patient_counts %>% filter(n_var < 30) %>% pull(n_var), probs = c(0.05,0.1, .25,.5,.75, .90, 0.95))
((plot1_histogram = plot_grid(hist, p, align = "v", nrow = 2)))
ggsave("ggsave/plot1_histogram.pdf", plot = plot1_histogram, width = 4, height = 4)
```

