---
title: "R Notebook"
output: html_notebook
---

What are the most frequent minor variants in the high-confidence samples?
```{r}
source("./scripts/startup.R")

consensus_sites_lowct<-fread("consensus_minor_changes_20220713.csv", data.table = F) %>% 
  mutate(MCoVNumber = mcov_reformat(name)) %>% filter(MCoVNumber %in% patient_var_30$MCoVNumber) %>% 
  filter(major != refnt)

consensus_sites_lowct
```


```{r}
# Across genome
patient_counts_30 = read_feather("processing/patient_counts_30.arrow")
consensus = consensus_sites_lowct %>% select(MCoVNumber, ntpos) %>% 
  group_by(ntpos) %>%
  tally(ntpos) %>% 
  mutate(consensus = ifelse(n > nrow(patient_counts_30 ) * 0.02, TRUE, FALSE)) %>%
  select(ntpos, consensus)

minor_consensus_plotdata = patient_var_30 %>% left_join(consensus) %>%
    mutate(consensus = replace_na(consensus, FALSE))

minor_prevalence_across_genome_unlog <- minor_consensus_plotdata %>% 
    filter(consensus==TRUE) %>% ggplot(aes(x=ntpos)) + 
#  scale_y_continuous(trans = "log2") +
  geom_bar(color = "grey") +

    # label your data too for the SALMON
    geom_bar(data = minor_consensus_plotdata %>% filter(consensus == FALSE), 
       aes(x = ntpos), color = "salmon") +
  # geom_point(data = minor_consensus_plotdata %>% filter(consensus == FALSE) %>% 
  #              group_by(ntpos) %>% tally(), 
  #          aes(x = ntpos, y= n), shape = "", color = "salmon") + 
    
    # label your data too for the GREY
    
      geom_point(data = minor_consensus_plotdata %>%
               group_by(ntpos) %>% tally(), 
           aes(x = ntpos, y= n), shape = "") + 
    scale_color_manual(values = c("salmon","grey")) + 
  theme_bw() + 
  geom_hline(yintercept = nrow(patient_counts_30)*0.02, linetype="dotted") + 
  xlab("Nucleotide position") + ylab("No. samples w/minor variant at site") + 
  annotate("rect", xmin=27894, xmax=28295, ymin=0, ymax=Inf, alpha=0.2, fill="#85D4E3") + 
  annotate("rect", xmin=28274, xmax=29533, ymin=0, ymax=Inf, alpha=0.2, fill="#F4B5BD") + 
  annotate("rect", xmin=21563, xmax=25384, ymin=0, ymax=Inf, alpha=0.2, fill="#FAD77B") + 
  xlim(0,29903)

minor_prevalence_across_genome_unlog_labeled = 
  minor_prevalence_across_genome_unlog + 
  geom_text_repel(ylim  = c(500,NA), size = 3, angle = 90, segment.linetype = 3,
                    segment.size = 0.6, segment.curvature = -1e-20, 
                    arrow = arrow(length = unit(0.015, "npc"), type = "closed"),
                    data = minor_consensus_plotdata %>% group_by(ntpos) %>% 
                      tally() %>% filter(n > nrow(patient_counts_30)*0.02) %>% 
                      left_join(minor_consensus_plotdata %>% select(ntpos, consensus)) %>% unique,
                    aes(x = ntpos, y = n, label = ntpos, color = consensus))

gene_limits

gene_arrows<-ggplot(gene_limits, aes(xmin = start, xmax = end, y=molecule, label = gene_id)) +
  geom_gene_arrow(arrowhead_height = unit(3, "mm"), arrowhead_width = unit(1, "mm")) + geom_gene_label() + geom_segment(aes(x=266,y=1.5,xend=13468,yend=1.5), size=0.2) + xlim(0,29903) +
  annotate("text", label="ORF1a", x=5000, y=1.41, size=3) + geom_segment(aes(x=13468,y=1.4,xend=21555,yend=1.4), size=0.2) + annotate("text", label="ORF1b", x=18000, y=1.31, size=3) +
  theme_genes() + ylab(NULL) + xlab(NULL) 



((
  genome_fig_unlog<-plot_grid(gene_arrows, minor_prevalence_across_genome_unlog_labeled, nrow=2, rel_heights=c(0.25,1), axis="lr", align="hv")
))
#ggsave("ggsave/genome_fig_unlog.pdf", plot = genome_fig_unlog, height = 5, width = 8)



# LOG TRANSFORM

((
  minor_prevalence_across_genome = minor_prevalence_across_genome_unlog + scale_y_continuous(trans = "log2") + geom_label_repel(size = 2, alpha = 0.65, data = minor_consensus_plotdata %>% 
               group_by(ntpos) %>% tally() %>% filter(n > nrow(patient_counts_30)*0.02) %>% 
                 left_join(minor_consensus_plotdata %>% select(ntpos, consensus)) %>% unique,
               aes(x = ntpos, y = n, label = ntpos, color = consensus))
))


# consensus_prevalence_across_genome<-consensus_sites_lowct %>% ggplot(aes(x=ntpos)) + 
#   geom_bar(fill="black", color="black") + theme_bw() + 
#   geom_hline(yintercept = nrow(patient_counts_30)*0.02, linetype="dotted") + 
#   xlab("Nucleotide position") + ylab("No. samples w/consensus variant at site") + 
#   annotate("rect", xmin=27894, xmax=28295, ymin=0, ymax=Inf, alpha=0.2, fill="#85D4E3") + 
#   annotate("rect", xmin=28274, xmax=29533, ymin=0, ymax=Inf, alpha=0.2, fill="#F4B5BD") + 
#   annotate("rect", xmin=21563, xmax=25384, ymin=0, ymax=Inf, alpha=0.2, fill="#FAD77B") + 
#   xlim(0,29903)

# prevalence_across_genome<-plot_grid(minor_prevalence_across_genome, consensus_prevalence_across_genome, nrow=2, axis="bltr")
# prevalence_across_genome

gene_limits<-genes %>% group_by(gene_id) %>% summarise(start=min(ntpos), end=max(ntpos)) %>% filter(gene_id!="") %>% mutate(molecule="")


((
  genome_fig<-plot_grid(gene_arrows, minor_prevalence_across_genome, nrow=2, rel_heights=c(0.25,1), axis="lr", align="hv")
))
#ggsave("ggsave/genome_fig.pdf", plot = genome_fig, height = 5, width = 8)




```

```{r}
#Frequency of mutations by gene in hospitalized vs. non-hospitalized patients
# gene_lengths <- genes %>% group_by(gene_id) %>% summarise(gene_length=n())
# 
# variant_density <- patient_var_30 %>% left_join(genes) %>% 
#   group_by(admitted_hospital, gene_id) %>% filter(gene_id!="") %>% 
#   summarise(n_variants=n()) %>% left_join(gene_lengths) %>% 
#   mutate(variant_density=n_variants/gene_length) 
# 
# 
# variant_density$gene_id<-factor(variant_density$gene_id, levels=c("5'UTR","nsp1","nsp2","PL-PRO","nsp4","3CL-PRO","nsp6","nsp7","nsp8","nsp9", "nsp10","RdRp","Hel","ExoN","NendoU","2'-O-MT","S","ORF3a","E","M","ORF6","ORF7ab","ORF8","N","ORF10","3'UTR"))
# #SUPP FIG 7
# ((
#   var_by_gene_fig_hosp<-variant_density %>% ggplot(aes(x=gene_id, y=variant_density)) + 
#     geom_bar(stat="identity", fill="white", color="black") + theme_bw() + 
#     theme(axis.text.x = element_text(angle=90)) + xlab("Gene") + ylab("Minor variant density") + 
#     facet_grid(admitted_hospital~.)
# ))
```


## Characteristics of highly recurrent minor variants

```{r}
((
  highly_shared_sites<- patient_var_30 %>% group_by(ntpos) %>% tally() %>% arrange(desc(n)) %>% 
    filter(n >= length(patient_var_30$name %>% unique)*0.02) %>% pull(ntpos) %>% na.omit() %>% as.numeric()
))


length(highly_shared_sites)
```

```{r}
# What's the range of minor allele frequencies the highly recurrent minor variants are found at?
((
  hss_maf<-patient_var_30 %>% filter(ntpos %in% highly_shared_sites) %>% 
    ggplot(aes(x = as.factor(ntpos), y = minorfreq)) + 
    geom_point(alpha = 0.2, position = position_jitter(width = 0.1)) + theme_bw() + 
    geom_boxplot(outlier.shape = NA, alpha =0.5) + 
    theme(axis.text.x = element_text(angle=90)) + #geom_boxplot(outlier.alpha = 0) + 
    xlab("Nucleotide position") + ylab("Minor allele freq") + 
    scale_y_continuous(trans="log2", breaks = c(0.01*2^(0:5), 0.5))
  ))

#ggsave("ggsave/hss_maf.pdf", width = 4, height = 2, plot = hss_maf)

((
  hss_maf_unlog<-patient_var_30 %>% filter(ntpos %in% highly_shared_sites) %>% 
    ggplot(aes(x = as.factor(ntpos), y = minorfreq)) + 
    geom_point(alpha = 0.2, position = position_jitter(width = 0.1)) + theme_bw() + 
    geom_boxplot(outlier.shape = NA, alpha =0.5) + 
    theme(axis.text.x = element_text(angle=90)) + #geom_boxplot(outlier.alpha = 0) + 
    xlab("Nucleotide position") + ylab("Minor allele freq") #+ 
    #scale_y_continuous(trans="log2", breaks = c(0.01*2^(0:5), 0.5))
  ))
#ggsave("ggsave/hss_maf_unlog.pdf", width = 4, height = 2, plot = hss_maf_unlog)

```


```{r, fig.height = 10}
#How many different changes do you find at each recurrently-mutated position? How many such mutations are a reversion to the reference?
hss_info <- fread('hss_data2.csv', data.table = F)

((
  hss_plot = patient_var_30 %>% filter(ntpos %in% highly_shared_sites) %>% 
    group_by(ntpos, major, minor) %>% summarise(n=n(), median_frequency = median(minorfreq)) %>% 
    left_join(hss_info) %>% mutate(change=paste0(major,">",minor)) %>% 
    mutate(ref_reversion=if_else(minor==refnt,"yes","no")) %>% 
    mutate(median_MAF=if_else(median_frequency<0.02, 
                              "low (median <2% MAF)","high (median >=2% MAF)")) %>%
    ggplot(aes(x=change, y=n, color=ref_reversion)) + 
    geom_bar(stat="identity", aes(fill=median_MAF)) + 
    scale_color_manual(values=c("gray","red")) + 
    scale_fill_manual(values=c("darkmagenta","steelblue")) + 
    facet_wrap(ntpos~`aa position`, scales="free") + 
    theme_pubr() + theme(axis.text.x=element_text(angle=90))
))

#ggsave('#ggsave/hss_plot.pdf', plot = hss_plot, width = 8, height = 10)
```



```{r, fig.height = 8, fig.width = 8}
#is appearance of these sites consistent over time?
# patient_var_30 %>% filter(ntpos %in% highly_shared_sites) %>% ggplot(aes(COLLECTION_DT)) + geom_bar() + scale_y_continuous(trans = "log1p", breaks = c(0,2^(0:5))) + facet_wrap(~ntpos) + theme_pubr() + theme(axis.text.x=element_text(angle=90), axis.text.y=element_text(size = 8))
```
## Frequency of minor variants associated with fitness increases

```{r}
mutations_of_interest <- fread('mutation_list.csv', data.table=F)
mutations_present_1 <- patient_var_30 %>% filter(ntpos %in% mutations_of_interest$ntpos) %>% left_join(mutations_of_interest) %>% select(MCoVNumber, COLLECTION_DT, ntpos, major, minor, mutation_name, allele, no_independent_emergences, notes) %>% filter(notes!="alternatively G") %>% mutate(mutation_present=if_else(minor==allele, 1,0)) %>% group_by(mutation_name, COLLECTION_DT) %>% summarise(n=sum(mutation_present)) 
mutations_present_2<-patient_var_30 %>% filter(ntpos %in% mutations_of_interest$ntpos) %>% left_join(mutations_of_interest) %>% select(MCoVNumber, COLLECTION_DT, ntpos, major, minor, mutation_name, allele, no_independent_emergences, notes) %>% filter(notes=="alternatively G") %>% mutate(mutation_present=if_else(minor=="A"|minor=="G", 1,0))%>% group_by(mutation_name, COLLECTION_DT) %>% summarise(n=sum(mutation_present)) 
mutations_present<-rbind(mutations_present_1, mutations_present_2)

setdiff(mutations_of_interest$mutation_name, mutations_present$mutation_name)
```

```{r}
#SUPP FIG 11
obermeyer = mutations_present %>% left_join(mutations_of_interest) %>% ggplot(aes(x=COLLECTION_DT, y=n)) + geom_bar(stat="identity", fill="black", color="black") + facet_wrap(mutation_name~ntpos) + theme_pubr() + theme(axis.text.x=element_text(angle=90), strip.background = element_rect(fill="lightblue")) + xlab("Collection Date") + ylab("No. samples") + scale_x_date(minor_breaks="1 month")

#ggsave("ggsave/obermeyer.pdf", plot = obermeyer, width = 6, height = 5)

```


```{r}
intersect(highly_shared_sites, mutations_of_interest$ntpos)
#One of the highly shared minor variants (corresponding to S: T95I) is associated with increased transmissibility
```

```{r}
#Do the highly shared sites tend to also be reproducible/highly shared? minors_table from reproducibility check in 01_load_new_samples.Rmd
minors_table %>% filter(detected_minor_in_repl=="yes") %>% filter(ntpos %in% highly_shared_sites) %>% group_by(ntpos) %>% tally() %>% arrange(desc(n))
#some of them are reproducibly present in >20% of samples
```

```{r}
consensus_hss<-fread('all_samples_hss.csv', data.table = F) %>% mutate(MCoVNumber=mcov_reformat(name)) %>% left_join(runs_all_samples)

empty_frame<-patient_var_30 %>% group_by(run, major) %>% tally() %>% 
  select(run, nt=major) %>% filter(nt != "")
consensus_minor_hss<-data.frame(run=as.character(), nt=as.character(), n=as.numeric(), type=as.character(), ntpos=as.numeric())

for (mutation in highly_shared_sites) {
  minors_site<-patient_var_30 %>% filter(ntpos==mutation) %>% 
    group_by(run, minor) %>% tally() %>% select(run, nt=minor, n) %>% 
    left_join(empty_frame) %>% mutate(n=replace_na(n, 0))
  minors_site$type<-"minor"
  minors_site$ntpos<-mutation
  consensus_site<-consensus_hss %>% filter(ntpos==mutation) %>% 
    filter(major %in% c("A","C","T","G")) %>% 
    group_by(run, major) %>% tally() %>% select(run, nt=major, n) %>% 
    left_join(empty_frame) %>% mutate(n=replace_na(n, 0))
  consensus_site$type<-"consensus"
  consensus_site$ntpos<-mutation
consensus_minor_hss<-rbind(consensus_minor_hss, minors_site, consensus_site)
}
```



```{r, fig.width = 8, fig.height = 8}
hss_plots = function(hss_1) {
    hss_1_plots = consensus_minor_hss %>% 
      mutate(run = str_remove(run, "Run_") %>% as.numeric) %>% 
      na.omit() %>% filter(ntpos %in% hss_1) %>% 
    ggplot(aes(x=run, y=n, fill=nt)) + geom_bar(stat="identity") +
    scale_fill_manual(values=viridis(5)) + facet_grid(type~ntpos, scales="free_y") + 
    theme_pubr() + theme(axis.text.x=element_text(angle=90), legend.position="none") +
      scale_x_continuous(breaks = seq(0,100,10))
    return(hss_1_plots)
}

# order highly_shared_sites by the y axis scale
hss_ordered = consensus_minor_hss %>% filter(type == "minor") %>% group_by(ntpos) %>%
  summarize(max_counts = max(n)) %>% arrange(max_counts) %>% 
  pull(ntpos) %>% unique()

hss_1_plots<-hss_plots(hss_ordered[1:5])
hss_2_plots<-hss_plots(hss_ordered[6:10])
hss_3_plots<-hss_plots(hss_ordered[11:15])
hss_4_plots = hss_plots(hss_ordered[16:20])
hss_5_plots = hss_plots(hss_ordered[21:length(hss_ordered)])

hss_plot_basepairs = plot_grid(hss_1_plots, hss_2_plots, hss_3_plots, hss_4_plots, 
                               hss_5_plots, nrow=5)

#ggsave("ggsave/hss_plot_basepairs.pdf", hss_plot_basepairs, height = 12, width = 10)
```

```{r}
#all minor variants that are found in at least 2% of high-coverage samples in Lythgoe et al https://www.science.org/doi/suppl/10.1126/science.abg0821/suppl_file/abg0821-lythgoe-sm.pdf
lythgoe_hss<-c(29320, 25628, 25807, 20364, 357, 25627, 25532, 11083, 239, 238, 356, 16740, 20989, 19393, 15009, 21826, 19937, 26289,28253,25507,29862,11052,29864,25529,369,22453,13571,29538,21575,75,20993,19473,15474,635,1226,29747,24933,19210,13303,28936,26334,16887)

tonkin_hss<-c(11075L, 11083L, 11074L, 522L, 26780L, 1547L, 14408L, 28253L, 13914L, 26137L, 635L, 241L, 26785L, 683L, 23403L, 28881L, 13778L, 25521L, 14805L, 9491L, 11071L, 13780L, 21575L, 29242L, 9430L, 203L, 3037L, 686L, 3096L, 16375L, 16466L, 17167L, 23559L, 9438L, 21137L, 29051L, 23555L, 26333L, 9474L, 13571L, 16887L, 26787L, 28603L, 24213L, 11651L, 27925L, 514L, 26781L, 558L, 1912L, 9479L, 11073L, 12578L, 26144L, 26681L, 29241L)
intersect(lythgoe_hss, highly_shared_sites)
intersect(tonkin_hss, highly_shared_sites)
```

```{r}
homoplasic_site_list<-c(187L, 1059L, 2094L, 3037L, 3130L, 6990L, 8022L, 10323L, 10741L, 11074L, 11083L, 13408L, 14786L, 15324L, 19684L, 20148L, 21137L, 21575L, 24034L, 24378L, 25563L, 26144L, 26461L, 26681L, 28077L, 28826L, 28854L, 29700L)
intersect(homoplasic_site_list, highly_shared_sites)
```

```{r}
tally_in_minors<-patient_var_30 %>% filter(ntpos %in% highly_shared_sites) %>% group_by(ntpos) %>% summarise(n_minor_variants_houston=n()) %>% mutate(freq_minor_variants_houston=n_minor_variants_houston/nrow(patient_counts_30))

tally_in_gisaid<-fread('hss_gisaid_tallies.csv', data.table = F) %>% select(ntpos, n_consensus_variants_gisaid=n) %>% mutate(freq_consensus_variants_gisaid=n_consensus_variants_gisaid/3109421) #USA gisaid sequences as of 6/13/2022
#SUPP FIG 10
((
  gisaid_vs_hmh_plot = tally_in_minors %>% left_join( tally_in_gisaid) %>% 
    ggplot(aes(x=freq_minor_variants_houston, y=freq_consensus_variants_gisaid, 
               color = as.factor(ntpos))) + geom_point(alpha=0.5) + theme_pubr(legend= "right") + 
  xlab("Fraction of samples with minor variant in Houston data") + 
    ylab("Fraction of samples with consensus variant in GISAID USA") + 
    geom_text_repel(aes(label=ntpos), max.overlaps = 10, label.size=0.1) + 
    geom_abline(slope = 1, intercept = 0, linetype="dashed", color = "gray")
))

#ggsave("ggsave/gisaid_vs_hmh_plot.pdf", plot = gisaid_vs_hmh_plot, width = 6, height = 5)
```

```{r}
# MUTATION SPECTRA ANALYSIS

# reproducibility analysis from RMD 01
plot_nonreproducible_spectra_heatmap = readRDS(file = "plot_nonreproducible_spectra_heatmap.rds")
plot_reproducible_spectra_heatmap = readRDS(file = "plot_reproducible_spectra_heatmap.rds")

hmap<-table(patient_var_30$major[patient_var_30$major!=""], patient_var_30$minor[patient_var_30$minor!=""]) %>% data.frame() %>%
  ggplot(aes(x=Var1, y=Var2, fill=Freq/sum(Freq))) + geom_tile(colour = "black") +
  scale_fill_gradient(low = "white",
                      high = "darkblue") +
  theme_minimal() +   labs(fill = "Number",
       x = "Consensus allele",
       y = "Minor allele", title = "Post-sample filter (Ct, n_var)") #Most frequently C>T mutations

# Now beecause some peaks are really prevalent like the twin peaks at 29187/8, 
# it's important to not factor that in too much. So let's just assign those mutations
# to equate to the median number a mutation appears which is 1.

mutation_spectra = patient_var_30 %>% unite(col = "spectra_mutation", 
                                            major, ntpos, minor, remove = FALSE)
mutation_spectra_counts = mutation_spectra %>% count(spectra_mutation)

mutation_spectra_counts %>% ggplot(aes(y=n)) + geom_boxplot() + 
  scale_y_continuous(trans = "log1p")

mutation_spectra_counts$n %>% quantile() # median is 1

mutation_spectra_unique = mutation_spectra %>% 
  select(major, minor, spectra_mutation) %>% unique()

hmap_unique <-table(mutation_spectra_unique$major[mutation_spectra_unique$major!=""], mutation_spectra_unique$minor[mutation_spectra_unique$minor!=""]) %>% data.frame() %>%
  ggplot(aes(x=Var1, y=Var2, fill=Freq/sum(Freq))) + geom_tile(colour = "black") + # grid color
  scale_fill_gradient(limits = c(0,0.4), low = "white",
                      high = "darkblue") +
  theme_minimal() +   labs(fill = "Fraction",
       x = "Consensus allele",
       y = "Minor allele", title="Unique minor variants")

rescale_fill = function(hmap_unique) {
  hmap_unique_rescaled = hmap_unique + 
    scale_fill_gradient(limits = c(0,0.4), low = "white",
                      high = "darkblue") +
    geom_label(fill = "white", alpha = 0.3, aes(x = Var1, y = Var2, label = round(Freq/sum(Freq), digits = 2)))
  return(hmap_unique_rescaled)
}

((
  rep_nucleotides = ggarrange(rescale_fill(plot_nonreproducible_spectra_heatmap), 
                              rescale_fill(plot_reproducible_spectra_heatmap), 
                              rescale_fill(hmap), 
                              rescale_fill(hmap_unique), common.legend = T)
))
rep_nucleotides

#ggsave("ggsave/heatmap_spectra_replicate_variants.pdf", rep_nucleotides, height = 6, width = 6)
```
