---
title: "R Notebook"
output: html_notebook
---

# Effects on patient analysis of using different detection thresholds
## Patient data - used for all analyses

# FUNCTION OF COVERAGE: Higher coverage to top 75%
# FUNCTION OF MAF: MAF increased to top 75%
# FUNCTION OF TITER: CT < 20 to top 75%

# FUNCTION OF QUADRANTS: CT > 35; Counts > 100

```{r}
source("./scripts/startup.R")
source("./scripts/load_data.R")
```

## Alternate Dataset 1: 2% MAF, 200x coverage, min. 100 reads to call minor variant
# find highest coverage

```{r}
#load file and tally minor variant richness
minor_variant_sites_threshold<-fread('alt_datasets/minor_variants_filtered_200x0.02_100.csv', data.table = F) 
mcov_samples_filtered<-mcov_samples %>% filter(!run %in% runs_to_drop) %>% 
  filter(qc_status=="pass") %>% filter(!MCoVNumber %in% nextclade_bad_samples) %>% filter(scorpio_call!="Omicron (BA.1-like)") %>% 
 ##### #main coverage criterion for fair comparisons: X depth over Y percent of the genome
  filter(fraction_200x_coverage>=0.98) %>% droplevels()
n_var<-minor_variant_sites_threshold %>% group_by(MCoVNumber) %>% tally() 
samples_n_var<-mcov_samples_filtered %>% left_join(n_var) %>% arrange(COLLECTION_DT) %>% mutate(n_var=replace_na(n, 0))
for_patient_analysis<-samples_n_var %>% filter(INSTRUMENT_RESULT<26)  %>% select(MCoVNumber, n_var, scorpio_call) 
for_patient_analysis %>% pull(n_var) %>% summary
```

```{r}
for_patient_analysis %>% nrow() 
median_var<-for_patient_analysis %>% pull(n_var) %>% median()
```

```{r}
#add patient metadata to sample minor variant richness
p<-left_join(for_patient_analysis, patient_data) %>% droplevels() %>% mutate(vocAlpha=if_else(startsWith(scorpio_call, "Alpha"),1,0), vocDelta=if_else(startsWith(scorpio_call, "Delta"),1,0)) %>% mutate(vocAlpha=as.factor(vocAlpha), vocDelta=as.factor(vocDelta))
```


### Association between patient hospitalization and high minor variant richness

```{r}
categories<-p %>% mutate(minor_greater_0=if_else(n_var>0,'yes','no')) %>% mutate(minor_greater_5=if_else(n_var>5,'yes','no')) %>% mutate(minor_greater_10=if_else(n_var>10,'yes','no')) 
categories1<-categories %>% group_by(admitted_hospital, minor_greater_0) %>% tally() %>% ggplot(aes(x=admitted_hospital, y=n, fill=minor_greater_0)) + geom_bar(stat="identity") + scale_fill_manual(values=c("gray","black")) + theme_bw()
categories2<-categories %>% group_by(admitted_hospital, minor_greater_5) %>% tally() %>% ggplot(aes(x=admitted_hospital, y=n, fill=minor_greater_5)) + geom_bar(stat="identity") + scale_fill_manual(values=c("gray","black")) + theme_bw()
categories3<-categories %>% group_by(admitted_hospital, minor_greater_10) %>% tally() %>% ggplot(aes(x=admitted_hospital, y=n, fill=minor_greater_10)) + geom_bar(stat="identity") + scale_fill_manual(values=c("gray","black")) + theme_bw()
plot_grid(categories1, categories2, categories3, nrow=3)
```

```{r}
hospitalization<-table(p$admitted_hospital,p$var_level) 
hospitalization
```

```{r}
chisq.test(hospitalization)
```

```{r}
oddsratio(hospitalization, rev="columns")
```

```{r}
#linear mixed-effect model with sequencing run as random effect
lme(log10(n_var+1) ~ CT*admitted_hospital, random=~1|run,
            data=p) %>% anova() 
```

```{r}
hosp_altdata1<-p %>% ggplot(aes(x=CT, y=log10(n_var+1), color=admitted_hospital)) + geom_point(alpha=0.5) + scale_color_manual(values=c("black","darkred")) + geom_smooth(method=lm) + theme_bw() + annotate("text", x=10, y=2, label="Ct p<0.0001 \nhospitalization p<0.0001 \nCt*hospitalization p<0.0001") + theme(legend.position="bottom") + ylim(0,2.6)
```



### Other ways of examining disease severity: vaccination status and healthcare worker surveillance

```{r}
#when did most cases among vaccinated patients occur?
table(p$collection_month, p$vaccine_status)
```


```{r}
vax_status_subset<- samples_n_var %>% filter(INSTRUMENT_RESULT<35)  %>% select(MCoVNumber, n_var, scorpio_call) %>% left_join(patient_data) %>% 
  filter(collection_date>="2021-07-01") %>% filter(admitted_hospital==0) 
lme(log10(n_var+1) ~ CT*vaccine_status, random=~1|run,
            data=vax_status_subset) %>% anova() 

```


```{r}
#to see effect of vaccination (as a correlate of disease severity): limit to later than July and only non-hospitalized patients
vax_altdata1<-vax_status_subset %>% ggplot(aes(x=CT, y=log10(n_var+1), color=vaccine_status)) + geom_point(alpha=0.5) + scale_color_manual(values=c("black","lightblue")) + geom_smooth(method=lm) + theme_bw() + annotate("text", x=14, y=2, label="Ct p<0.0001 \nvaccination p=0.0119 \nCt*vaccination p=0.055") + theme(legend.position="bottom") + ylim(0,2.6)
```

```{r}
#healthcare worker surveillance (presumed mostly asymptomatic) vs non-hospitalized patients (presumed mostly symptomatic)
surv_subset <- samples_n_var %>% filter(INSTRUMENT_RESULT<35)  %>% select(MCoVNumber, n_var, scorpio_call) %>% left_join(patient_data) %>%
  filter(admitted_hospital==0) %>% filter(!(surveillance_sample==1 & pui=="PUI")) #exclude HCW who were patients
lme(log10(n_var+1) ~ CT*surveillance_sample, random=~1|run,
            data=surv_subset) %>% anova() 
```

```{r}
hcw_altdata1<-surv_subset %>% ggplot(aes(x=CT, y=log10(n_var+1), color=surveillance_sample)) + geom_point(alpha=0.5) + scale_color_manual(values=c("black","darkgreen")) + geom_smooth(method=lm) + theme_bw() + annotate("text", x=12, y=2, label="Ct p<0.0001 \nHCW surveillance p=0.007 \nCt*surveillance p=0.0076")+ theme(legend.position="bottom") + ylim(0,2.6)
```

```{r}
#LASSO regression model including all sample and patient characteristics to explain minor variant richness
p_sub<- p %>% left_join(mcov_samples) #to add info about coverage
y<-log10(p_sub$n_var+1)
x<-data.matrix(p_sub[, c('age18under','age55plus','sex','chronic_lung_disease', 'chronic_liver_disease', 'chronic_kidney_disease', 'chronic_heart_disease', 'hypertension', 'diabetes', 'cancer', 'obesity', 'plasma', 'mAb', 'admitted_hospital','vaccine_status','vocAlpha','vocDelta','collection_month','surveillance_sample','CT','median_coverage','run')])

cv_model <- cv.glmnet(x, y, alpha = 1, nfolds=100)
plot(cv_model) 
best_model <- glmnet(x, y, alpha = 1, lambda = cv_model$lambda.min)
best_model$dev.ratio
```

```{r}
lasso1_altdata1<-coef(best_model) %>% as.matrix() %>% data.frame() %>% rownames_to_column() %>% rename(factor=rowname, coefficient=s0) %>% filter(factor!="(Intercept)") %>% arrange(desc(coefficient)) %>% ggplot(aes(x=coefficient, y=fct_reorder(factor,coefficient))) + geom_bar(stat="identity", fill="#E2D200", color="gray") + theme_bw() + ylab("Factor") + xlab("Coefficient")
```

```{r}
#same model excluding any factors with a temporal signal (VOC, vaccination, collection month, run)
y2<-log10(p_sub$n_var+1)
x2<-data.matrix(p_sub[, c('age18under','age55plus','sex','chronic_lung_disease', 'chronic_liver_disease', 'chronic_kidney_disease', 'chronic_heart_disease', 'hypertension', 'diabetes', 'cancer', 'obesity', 'plasma', 'mAb', 'admitted_hospital', 'surveillance_sample', 'CT', 'median_coverage')])

cv_model_2 <- cv.glmnet(x2, y2, alpha = 1, nfolds=100)
plot(cv_model_2) 
best_model_2 <- glmnet(x2, y2, alpha = 1, lambda = cv_model$lambda.min)
best_model_2$dev.ratio
```

```{r}
lasso2_altdata1<-coef(best_model_2) %>% as.matrix() %>% data.frame() %>% rownames_to_column() %>% rename(factor=rowname, coefficient=s0) %>% filter(factor!="(Intercept)") %>% arrange(desc(coefficient)) %>% ggplot(aes(x=coefficient, y=fct_reorder(factor,coefficient))) + geom_bar(stat="identity", fill="#E2D200", color="black") + theme_bw() + ylab("Factor") + xlab("Coefficient")
```

```{r}
title <- ggdraw() + 
  draw_label(
    "Alternate dataset 1",
    fontface = 'bold',
    x = 0,
    hjust = 0)

all_plots_altdata1<-plot_grid(plot_grid(hosp_altdata1, vax_altdata1, hcw_altdata1, ncol=3), plot_grid(lasso1_altdata1, lasso2_altdata1, ncol=2), nrow=2, rel_heights = c(2,1))

alt_1<-plot_grid(title, all_plots_altdata1, nrow=2, rel_heights=c(0.1,1))
```


## Alternate Dataset 2: 3% MAF, min 500x coverage, min 20 reads

```{r}
#load file and tally minor variant richness
minor_variant_sites_threshold<-fread('alt_datasets/minor_variants_filtered_500x0.03_20.csv', data.table = F) 
mcov_samples_filtered<-mcov_samples %>% filter(!run %in% runs_to_drop) %>% 
  filter(qc_status=="pass") %>% filter(!MCoVNumber %in% nextclade_bad_samples) %>% filter(scorpio_call!="Omicron (BA.1-like)") %>% 
 ##### #main coverage criterion for fair comparisons: X depth over Y percent of the genome
  filter(fraction_500x_coverage>=0.98) %>% droplevels()
n_var<-minor_variant_sites_threshold %>% group_by(MCoVNumber) %>% tally() 
samples_n_var<-mcov_samples_filtered %>% left_join(n_var) %>% arrange(COLLECTION_DT) %>% mutate(n_var=replace_na(n, 0))
for_patient_analysis<-samples_n_var %>% filter(INSTRUMENT_RESULT<26)  %>% select(MCoVNumber, n_var, scorpio_call) 
for_patient_analysis %>% pull(n_var) %>% summary

```

```{r}
for_patient_analysis %>% nrow() 
median_var<-for_patient_analysis %>% pull(n_var) %>% median()
```

```{r}
#add patient metadata to sample minor variant richness
p<-left_join(for_patient_analysis, patient_data) %>% droplevels() %>% mutate(vocAlpha=if_else(startsWith(scorpio_call, "Alpha"),1,0), vocDelta=if_else(startsWith(scorpio_call, "Delta"),1,0)) %>% mutate(vocAlpha=as.factor(vocAlpha), vocDelta=as.factor(vocDelta))
```

```{r}
#Random Forest model of patient characteristics for classifying low-CT samples as having high or low minor variant richness
p <- p %>% mutate(var_level=if_else(n_var<=median_var, "low","high")) %>% mutate(var_level=as.factor(var_level)) 

p_select <- p  %>% select(age18under, age18to54, age55plus, sex, ethnicity, chronic_lung_disease, chronic_liver_disease, chronic_kidney_disease, chronic_heart_disease, hypertension, diabetes, cancer, obesity, transplant_patient, plasma, mAb, vaccine_status, admitted_hospital,surveillance_sample, var_level) 

p.rf<-randomForest(var_level~.,
                               data=p_select, 
                               ntree=1000,
                                mtry=5,
                               importance = T) 

importance.randomForest <- as.data.frame(randomForest::importance(p.rf))

importance.randomForest<-importance.randomForest %>% arrange(desc(MeanDecreaseAccuracy))
importance.randomForest

rf.roc<-roc(p_select$var_level,p.rf$votes[,2], levels=c(case="high",control="low"))
auc(rf.roc)
```

### Association between patient hospitalization and high minor variant richness

```{r}
categories<-p %>% mutate(minor_greater_0=if_else(n_var>0,'yes','no')) %>% mutate(minor_greater_5=if_else(n_var>5,'yes','no')) %>% mutate(minor_greater_10=if_else(n_var>10,'yes','no')) 
categories1<-categories %>% group_by(admitted_hospital, minor_greater_0) %>% tally() %>% ggplot(aes(x=admitted_hospital, y=n, fill=minor_greater_0)) + geom_bar(stat="identity") + scale_fill_manual(values=c("gray","black")) + theme_bw()
categories2<-categories %>% group_by(admitted_hospital, minor_greater_5) %>% tally() %>% ggplot(aes(x=admitted_hospital, y=n, fill=minor_greater_5)) + geom_bar(stat="identity") + scale_fill_manual(values=c("gray","black")) + theme_bw()
categories3<-categories %>% group_by(admitted_hospital, minor_greater_10) %>% tally() %>% ggplot(aes(x=admitted_hospital, y=n, fill=minor_greater_10)) + geom_bar(stat="identity") + scale_fill_manual(values=c("gray","black")) + theme_bw()
plot_grid(categories1, categories2, categories3, nrow=3)
```

```{r}
hospitalization<-table(p$admitted_hospital,p$var_level) 
hospitalization
```

```{r}
chisq.test(hospitalization)
```

```{r}
oddsratio(hospitalization, rev="columns")
```

```{r}
#linear mixed-effect model with sequencing run as random effect
lme(log10(n_var+1) ~ CT*admitted_hospital, random=~1|run,
            data=p) %>% anova() 
```

```{r}
hosp_altdata2<-p %>% ggplot(aes(x=CT, y=log10(n_var+1), color=admitted_hospital)) + geom_point(alpha=0.5) + scale_color_manual(values=c("black","darkred")) + geom_smooth(method=lm) + theme_bw() + annotate("text", x=10, y=1.75, label="Ct p<0.0001 \nhospitalization p<0.0001 \nCt*hospitalization p=0.0059") + theme(legend.position="bottom") + ylim(0,2.6)
hosp_altdata2
```

```{r}
vax_status_subset<- samples_n_var %>% filter(INSTRUMENT_RESULT<35)  %>% select(MCoVNumber, n_var, scorpio_call) %>% left_join(patient_data) %>% 
  filter(collection_date>="2021-07-01") %>% filter(admitted_hospital==0) 
lme(log10(n_var+1) ~ CT*vaccine_status, random=~1|run,
            data=vax_status_subset) %>% anova() 
```

```{r}
#to see effect of vaccination (as a correlate of disease severity): limit to later than July and only non-hospitalized patients
vax_altdata2<-vax_status_subset %>% ggplot(aes(x=CT, y=log10(n_var+1), color=vaccine_status)) + geom_point(alpha=0.5) + scale_color_manual(values=c("black","lightblue")) + geom_smooth(method=lm) + theme_bw() + annotate("text", x=14, y=1.75, label="Ct p<0.0001 \nvaccination p=0.0671 \nCt*vaccination p=0.0226") + theme(legend.position="bottom") + ylim(0,2.6)
vax_altdata2
```

```{r}
#healthcare worker surveillance (presumed mostly asymptomatic) vs non-hospitalized patients (presumed mostly symptomatic)
surv_subset <- samples_n_var %>% filter(INSTRUMENT_RESULT<35)  %>% select(MCoVNumber, n_var, scorpio_call) %>% left_join(patient_data) %>%
  filter(admitted_hospital==0) %>% filter(!(surveillance_sample==1 & pui=="PUI")) #exclude HCW who were patients
lme(log10(n_var+1) ~ CT*surveillance_sample, random=~1|run,
            data=surv_subset) %>% anova() 
```

```{r}
hcw_altdata2<-surv_subset %>% ggplot(aes(x=CT, y=log10(n_var+1), color=surveillance_sample)) + geom_point(alpha=0.5) + scale_color_manual(values=c("black","darkgreen")) + geom_smooth(method=lm) + theme_bw() + annotate("text", x=12, y=1.75, label="Ct p<0.0001 \nHCW surveillance p=0.0077 \nCt*surveillance p=0.0001")+ theme(legend.position="bottom") + ylim(0,2.6)
hcw_altdata2
```

```{r}
#LASSO regression model including all sample and patient characteristics to explain minor variant richness
p_sub<- p %>% left_join(mcov_samples) #to add info about coverage
y<-log10(p_sub$n_var+1)
x<-data.matrix(p_sub[, c('age18under','age55plus','sex','chronic_lung_disease', 'chronic_liver_disease', 'chronic_kidney_disease', 'chronic_heart_disease', 'hypertension', 'diabetes', 'cancer', 'obesity', 'plasma', 'mAb', 'admitted_hospital','vaccine_status','vocAlpha','vocDelta','collection_month','surveillance_sample','CT','median_coverage','run')])

cv_model <- cv.glmnet(x, y, alpha = 1, nfolds=100)
plot(cv_model) 
best_model <- glmnet(x, y, alpha = 1, lambda = cv_model$lambda.min)
best_model$dev.ratio
```

```{r}
lasso1_altdata2<-coef(best_model) %>% as.matrix() %>% data.frame() %>% rownames_to_column() %>% rename(factor=rowname, coefficient=s0) %>% filter(factor!="(Intercept)") %>% arrange(desc(coefficient)) %>% ggplot(aes(x=coefficient, y=fct_reorder(factor,coefficient))) + geom_bar(stat="identity", fill="#E2D200", color="gray") + theme_bw() + ylab("Factor") + xlab("Coefficient")
```

```{r}
#same model excluding any factors with a temporal signal (VOC, vaccination, collection month, run)
y2<-log10(p_sub$n_var+1)
x2<-data.matrix(p_sub[, c('age18under','age55plus','sex','chronic_lung_disease', 'chronic_liver_disease', 'chronic_kidney_disease', 'chronic_heart_disease', 'hypertension', 'diabetes', 'cancer', 'obesity', 'plasma', 'mAb', 'admitted_hospital', 'surveillance_sample', 'CT', 'median_coverage')])

cv_model_2 <- cv.glmnet(x2, y2, alpha = 1, nfolds=100)
plot(cv_model_2) 
best_model_2 <- glmnet(x2, y2, alpha = 1, lambda = cv_model$lambda.min)
best_model_2$dev.ratio
```

```{r}
lasso2_altdata2<-coef(best_model_2) %>% as.matrix() %>% data.frame() %>% rownames_to_column() %>% rename(factor=rowname, coefficient=s0) %>% filter(factor!="(Intercept)") %>% arrange(desc(coefficient)) %>% ggplot(aes(x=coefficient, y=fct_reorder(factor,coefficient))) + geom_bar(stat="identity", fill="#E2D200", color="black") + theme_bw() + ylab("Factor") + xlab("Coefficient")
```

```{r}
title <- ggdraw() + 
  draw_label(
    "Alternate dataset 2",
    fontface = 'bold',
    x = 0,
    hjust = 0)

all_plots_altdata2<-plot_grid(plot_grid(hosp_altdata2, vax_altdata2, hcw_altdata2, ncol=3), plot_grid(lasso1_altdata2, lasso2_altdata2, ncol=2), nrow=2, rel_heights = c(2,1))

alt_2<-plot_grid(title, all_plots_altdata2, nrow=2, rel_heights=c(0.1,1))
alt_2
```



## Alternate Dataset 3: MAF 1%, minimum 1000x coverage 

```{r}
#load file and tally minor variant richness
minor_variant_sites_threshold<-fread('alt_datasets/minor_variants_filtered_1000x0.01.csv', data.table = F) 
mcov_samples_filtered<-mcov_samples %>% filter(!run %in% runs_to_drop) %>% 
  filter(qc_status=="pass") %>% filter(!MCoVNumber %in% nextclade_bad_samples) %>% filter(scorpio_call!="Omicron (BA.1-like)") %>% 
 ##### #main coverage criterion for fair comparisons: X depth over Y percent of the genome
  filter(fraction_1000x_coverage>=0.98) %>% droplevels()
n_var<-minor_variant_sites_threshold %>% group_by(MCoVNumber) %>% tally() 
samples_n_var<-mcov_samples_filtered %>% left_join(n_var) %>% arrange(COLLECTION_DT) %>% mutate(n_var=replace_na(n, 0))
for_patient_analysis<-samples_n_var %>% filter(INSTRUMENT_RESULT<26)  %>% select(MCoVNumber, n_var, scorpio_call) 
for_patient_analysis %>% pull(n_var) %>% summary
```

```{r}
for_patient_analysis %>% nrow() 
median_var<-for_patient_analysis %>% pull(n_var) %>% median()
```


```{r}
#add patient metadata to sample minor variant richness
p<-left_join(for_patient_analysis, patient_data) %>% droplevels() %>% mutate(vocAlpha=if_else(startsWith(scorpio_call, "Alpha"),1,0), vocDelta=if_else(startsWith(scorpio_call, "Delta"),1,0)) %>% mutate(vocAlpha=as.factor(vocAlpha), vocDelta=as.factor(vocDelta))
```

```{r}
#Random Forest model of patient characteristics for classifying low-CT samples as having high or low minor variant richness
p <- p %>% mutate(var_level=if_else(n_var<=median_var, "low","high")) %>% mutate(var_level=as.factor(var_level)) 

p_select <- p  %>% select(age18under, age18to54, age55plus, sex, ethnicity, chronic_lung_disease, chronic_liver_disease, chronic_kidney_disease, chronic_heart_disease, hypertension, diabetes, cancer, obesity, transplant_patient, plasma, mAb, vaccine_status, admitted_hospital,surveillance_sample, var_level) 

p.rf<-randomForest(var_level~.,
                               data=p_select, 
                               ntree=1000,
                                mtry=5,
                               importance = T) 

importance.randomForest <- as.data.frame(randomForest::importance(p.rf))

importance.randomForest<-importance.randomForest %>% arrange(desc(MeanDecreaseAccuracy))
importance.randomForest

rf.roc<-roc(p_select$var_level,p.rf$votes[,2], levels=c(case="high",control="low"))
auc(rf.roc)
```

### Association between patient hospitalization and high minor variant richness

```{r}
categories<-p %>% mutate(minor_greater_0=if_else(n_var>0,'yes','no')) %>% mutate(minor_greater_5=if_else(n_var>5,'yes','no')) %>% mutate(minor_greater_10=if_else(n_var>10,'yes','no')) 
categories1<-categories %>% group_by(admitted_hospital, minor_greater_0) %>% tally() %>% ggplot(aes(x=admitted_hospital, y=n, fill=minor_greater_0)) + geom_bar(stat="identity") + scale_fill_manual(values=c("gray","black")) + theme_bw()
categories2<-categories %>% group_by(admitted_hospital, minor_greater_5) %>% tally() %>% ggplot(aes(x=admitted_hospital, y=n, fill=minor_greater_5)) + geom_bar(stat="identity") + scale_fill_manual(values=c("gray","black")) + theme_bw()
categories3<-categories %>% group_by(admitted_hospital, minor_greater_10) %>% tally() %>% ggplot(aes(x=admitted_hospital, y=n, fill=minor_greater_10)) + geom_bar(stat="identity") + scale_fill_manual(values=c("gray","black")) + theme_bw()
plot_grid(categories1, categories2, categories3, nrow=3)
```

```{r}
hospitalization<-table(p$admitted_hospital,p$var_level) 
hospitalization
```

```{r}
chisq.test(hospitalization)
```

```{r}
oddsratio(hospitalization, rev="columns")
```

```{r}
#linear mixed-effect model with sequencing run as random effect
lme(log10(n_var+1) ~ CT*admitted_hospital, random=~1|run,
            data=p) %>% anova() 
```

```{r}
hosp_altdata3<-p %>% ggplot(aes(x=CT, y=log10(n_var+1), color=admitted_hospital)) + geom_point(alpha=0.5) + scale_color_manual(values=c("black","darkred")) + geom_smooth(method=lm) + theme_bw() + annotate("text", x=10, y=2, label="Ct p<0.0001 \nhospitalization p<0.0001 \nCt*hospitalization p<0.0001") + theme(legend.position="bottom") + ylim(0,2.6)
hosp_altdata3
```

```{r}
vax_status_subset<- samples_n_var %>% filter(INSTRUMENT_RESULT<35)  %>% select(MCoVNumber, n_var, scorpio_call) %>% left_join(patient_data) %>% 
  filter(collection_date>="2021-07-01") %>% filter(admitted_hospital==0) 
lme(log10(n_var+1) ~ CT*vaccine_status, random=~1|run,
            data=vax_status_subset) %>% anova() 

```

```{r}
#to see effect of vaccination (as a correlate of disease severity): limit to later than July and only non-hospitalized patients
vax_altdata3<-vax_status_subset %>% ggplot(aes(x=CT, y=log10(n_var+1), color=vaccine_status)) + geom_point(alpha=0.5) + scale_color_manual(values=c("black","lightblue")) + geom_smooth(method=lm) + theme_bw() + annotate("text", x=14, y=2, label="Ct p<0.0001 \nvaccination p=0.1098 \nCt*vaccination p=0.56") + theme(legend.position="bottom") + ylim(0,2.6)
vax_altdata3
```

```{r}
#healthcare worker surveillance (presumed mostly asymptomatic) vs non-hospitalized patients (presumed mostly symptomatic)
surv_subset <- samples_n_var %>% filter(INSTRUMENT_RESULT<35)  %>% select(MCoVNumber, n_var, scorpio_call) %>% left_join(patient_data) %>%
  filter(admitted_hospital==0) %>% filter(!(surveillance_sample==1 & pui=="PUI")) #exclude HCW who were patients
lme(log10(n_var+1) ~ CT*surveillance_sample, random=~1|run,
            data=surv_subset) %>% anova() 
```

```{r}
hcw_altdata3<-surv_subset %>% ggplot(aes(x=CT, y=log10(n_var+1), color=surveillance_sample)) + geom_point(alpha=0.5) + scale_color_manual(values=c("black","darkgreen")) + geom_smooth(method=lm) + theme_bw() + annotate("text", x=12, y=2, label="Ct p<0.0001 \nHCW surveillance p=0.449 \nCt*surveillance p=0.0019")+ theme(legend.position="bottom") + ylim(0,2.6)
hcw_altdata3
```

```{r}
#LASSO regression model including all sample and patient characteristics to explain minor variant richness
p_sub<- p %>% left_join(mcov_samples) #to add info about coverage
y<-log10(p_sub$n_var+1)
x<-data.matrix(p_sub[, c('age18under','age55plus','sex','chronic_lung_disease', 'chronic_liver_disease', 'chronic_kidney_disease', 'chronic_heart_disease', 'hypertension', 'diabetes', 'cancer', 'obesity', 'plasma', 'mAb', 'admitted_hospital','vaccine_status','vocAlpha','vocDelta','collection_month','surveillance_sample','CT','median_coverage','run')])

cv_model <- cv.glmnet(x, y, alpha = 1, nfolds=100)
plot(cv_model) 
best_model <- glmnet(x, y, alpha = 1, lambda = cv_model$lambda.min)
best_model$dev.ratio
```

```{r}
lasso1_altdata3<-coef(best_model) %>% as.matrix() %>% data.frame() %>% rownames_to_column() %>% rename(factor=rowname, coefficient=s0) %>% filter(factor!="(Intercept)") %>% arrange(desc(coefficient)) %>% ggplot(aes(x=coefficient, y=fct_reorder(factor,coefficient))) + geom_bar(stat="identity", fill="#E2D200", color="gray") + theme_bw() + ylab("Factor") + xlab("Coefficient")
```

```{r}
#same model excluding any factors with a temporal signal (VOC, vaccination, collection month, run)
y2<-log10(p_sub$n_var+1)
x2<-data.matrix(p_sub[, c('age18under','age55plus','sex','chronic_lung_disease', 'chronic_liver_disease', 'chronic_kidney_disease', 'chronic_heart_disease', 'hypertension', 'diabetes', 'cancer', 'obesity', 'plasma', 'mAb', 'admitted_hospital', 'surveillance_sample', 'CT', 'median_coverage')])

cv_model_2 <- cv.glmnet(x2, y2, alpha = 1, nfolds=100)
plot(cv_model_2) 
best_model_2 <- glmnet(x2, y2, alpha = 1, lambda = cv_model$lambda.min)
best_model_2$dev.ratio
```

```{r}
lasso2_altdata3<-coef(best_model_2) %>% as.matrix() %>% data.frame() %>% rownames_to_column() %>% rename(factor=rowname, coefficient=s0) %>% filter(factor!="(Intercept)") %>% arrange(desc(coefficient)) %>% ggplot(aes(x=coefficient, y=fct_reorder(factor,coefficient))) + geom_bar(stat="identity", fill="#E2D200", color="black") + theme_bw() + ylab("Factor") + xlab("Coefficient")
```

```{r}
title <- ggdraw() + 
  draw_label(
    "Alternate dataset 3",
    fontface = 'bold',
    x = 0,
    hjust = 0)

all_plots_altdata3<-plot_grid(plot_grid(hosp_altdata3, vax_altdata3, hcw_altdata3, ncol=3), plot_grid(lasso1_altdata3, lasso2_altdata3, ncol=2), nrow=2, rel_heights = c(2,1))

alt_3<-plot_grid(title, all_plots_altdata3, nrow=2, rel_heights=c(0.1,1))
alt_3
```


